Add tools/update-submodules.sh + a few other changes to how we build things

niklasdewally opened the pull request:
Closes #108 

@ChrisJefferson is this the kind of thing you had in mind?

ozgurakgun commented on clean.sh:
```diff
@@ -0,0 +1,11 @@
+cd $(dirname "$0")
+
+# clear submodules
+bash ./etc/scripts/get_submodule_paths.sh | xargs rm -rf
```
I guess this is fine, bu could also do something like `git clean -dfx .` in each submodule so we don't fetch a fresh copy?

niklasdewally commented:
@ozgurakgun I've done those changes.

I think git submodule sync then update covers all that could possibly change with submodules over time (i.e. what repo and branch it should point to, and updating it to new commits)?

ChrisJefferson commented:
This doesn't remove solvers/chuffed/libwrapper.a and solvers/chuffed/wrapper.o . Not entirely sure who/what is responsable for that? can we teach `cargo clean` to do that? or do we need to run ./clean.sh from solvers/chuffed?

ozgurakgun commented:
Here they say putting the build artefacts under OUT_DIR is the way to go: https://github.com/rust-lang/cargo/issues/2573

ozgurakgun commented:
And this make me think configurable cargo clean isn't a thing: https://github.com/rust-lang/cargo/issues/572

niklasdewally commented:
Found some typos in build.rs files! ( i think).

niklasdewally commented:
> > Here they say putting the build artefacts under OUT_DIR is the way to go: [rust-lang/cargo#2573](https://github.com/rust-lang/cargo/issues/2573)
> 
> I'll open an issue for this so that Kieran can look at it when he has time (alongside the build stuff chris mentioned in the other PR). From the cargo book, it seems that it is heavily recommended to only use OUT_DIR in build scripts (probably for these reasons).
> 
> It looks like it should pass down as an env var. https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-sets-for-build-scripts

As I was in the build.rs files anyways, I have fixed this.

Also added the rerun-if-changed for the wrapper stuff as in #106.

I think the typo i fixed for this and minion would explain some issues Felix was having the other day - cargo build should've rebuilt the stuff in vendor but it didn't due to the typo, so we got a linking error.

niklasdewally commented:
@Kieranoski702 As i've changed some chuffed stuff, would appreciate a review from you as well.

niklasdewally commented:
> I think the typo i fixed for this and minion would explain some issues Felix was having the other day - cargo build should've rebuilt the stuff in vendor but it didn't due to the typo, so we got a linking error.

@ChrisJefferson @ozgurakgun  my understanding is with this fixed in [873d0d4](https://github.com/conjure-cp/conjure-oxide/pull/109/commits/873d0d4a13beefcbd432bd895815a58c802f3e89) we shouldn't need to clean build things on vendor update, but it might be nice to have anyways?

niklasdewally commented:
> Here they say putting the build artefacts under OUT_DIR is the way to go: [rust-lang/cargo#2573](https://github.com/rust-lang/cargo/issues/2573)

Should we do this for solver build artifacts too (i.e the stuff currently in vendor/build), or is what we have good enough for now?

ozgurakgun commented on solvers/minion/build.sh:
```diff
@@ -2,20 +2,17 @@
+git submodule sync -- vendor 
+git submodule update --init --recursive -- vendor
+
 cd "$SCRIPT_DIR"
```
it looks like this cd wants to be before the git commands, am I missing something?

niklasdewally replied:
thank you - thats a typo on my part.

ozgurakgun commented on solvers/chuffed/build.sh:
```diff
@@ -1,17 +1,26 @@
 #!/bin/bash
 
-SCRIPT_DIR=$(dirname "$0")
+SCRIPT_DIR=$(realpath $(dirname "$0"))
```
why the realpath here and not in the minion one? also same question about location of cd.

niklasdewally replied:
this has already been fixed in a WIP commit - the short answer is that I am doing more stuff with minion atm.

ozgurakgun commented on etc/scripts/get_submodule_paths.sh:
```diff
@@ -6,7 +6,7 @@
 
 # go to top level of git repo
-cd $(git rev-parse --show-toplevel)
+cd $(git rev-parse --show-superproject-working-tree --show-toplevel | head -n 1)
```
I learned a new arg to git today! `--show-superproject-working-tree`

ozgurakgun commented:
Sorry I added comments one-by-one instead of a combined review. I always think I'll just say this one thing, and end up saying more things...

If we managed to make cargo clean/build do the right thing I vote for removing clean.sh.

niklasdewally commented:
> Sorry I added comments one-by-one instead of a combined review. I always think I'll just say this one thing, and end up saying more things...
> 
> If we managed to make cargo clean/build do the right thing I vote for removing clean.sh.

I have made it do the right thing, I think.

I'll leave clean.sh for now for Chris and Kieran to review, just in case I've missed some C++ stuff somewhere.

ozgurakgun reviewed (approved):
LGTM as far as anyone can review shell scripts by reading...

ChrisJefferson commented:
cargo clean now gets rid of all the C++ files, as far as I can see. I think having a global clean.sh to reset all the submodules is still nice (as it does seem to require multiple lines), but cargo clean will now do most of the work.

ozgurakgun commented:
So there are 3 parts to the clean script:

- `git submodule sync --recursive` my understanding is that this updates the submodules to use the up to date URLs from the .gitmodules file
- `git submodule update --init --recursive` this updates the submodules to their latest version
- ` git clean -dfx .` on submodules to remove untracked files etc

So this is less to do with cleaning and more to do with updating the submodules, right? Should we rename it to `update-submodules.sh`? Or even `etc/build/update-submodules.sh` so it's not sitting at the top?

We should also add a link to the readme file pointing to the [contributing page](https://github.com/conjure-cp/conjure-oxide/wiki/Contributing) which I see @niklasdewally updated yesterday to mention clean.sh.

niklasdewally commented:
> So there are 3 parts to the clean script:
> 
> - `git submodule sync --recursive` my understanding is that this updates the submodules to use the up to date URLs from the .gitmodules file
> - `git submodule update --init --recursive` this updates the submodules to their latest version
> - ` git clean -dfx .` on submodules to remove untracked files etc
> 
> So this is less to do with cleaning and more to do with updating the submodules, right? Should we rename it to `update-submodules.sh`? Or even `etc/build/update-submodules.sh` so it's not sitting at the top?
> 
> We should also add a link to the readme file pointing to the [contributing page](https://github.com/conjure-cp/conjure-oxide/wiki/Contributing) which I see @niklasdewally updated yesterday to mention clean.sh.

etc/scripts/update-submodules.sh sounds good


Would tools be a better name for the etc directory? It might be nicer if we have

tools/foo.sh instead of etc/scripts/foo.sh from an ergonomic perspective

ozgurakgun commented:
I am happy with that since tools exists already.

niklasdewally commented:
> I am happy with that since tools exists already.

Oh I didn't notice that! Consistent directory structure either way would be good and tools actually says what it does so.

niklasdewally commented:
I've just added some stuff to the wiki page about this and installation generally as well
